### ISSUE TITLE
Upgrade serialization of chainWork field of `StoredBlock` to support more than 12 bytes (96 bits)

### ISSUE THREAD

**Issue by msgilligan** [bitcoinj/bitcoinj#3415]
**NOTE: The short-term, critical issue that occurred on June 24, 2024 has been fixed** in the [0.16.3](https://github.com/bitcoinj/bitcoinj/releases/tag/v0.16.3) release, see Issue #3410

This issue is about being prepared for when chainwork doubles again (in roughly 1 year)

As of block 849,138 chainWork has reached a value of `800028897aa1dce163ead2cf` which is too big to store in 12 **signed** bytes, though it will fit in 12 **unsigned** bytes. This was reported in Issue #3410 and fixed by PR #3411 (which converted the serialization to output an unsigned value.)

This fix buys us some time, but we will need a longer-term situation.

A proper solution might be to add a version field at the start of each `StoredBlock` record -- or perhaps the start of the file or in some metadata somewhere.  It could be as simple as `0xFF01` indicating a new version. But we will need to do something before the chain work value reaches roughly double what it is now.

For now, I'm going to mark this as targeted for the 0.17 release, but we may be able to postpone until 0.18.

**Comment by ulserGit** [comment]
Good afternoon, my blockchain download process stopped at block 849137
I'm using BitcoinJ 16.2 from the Maven repository
How can I solve the problem?

**Comment by msgilligan** [comment]
> my blockchain download process stopped at block 849137 I'm using BitcoinJ 16.2 from the Maven repository How can I solve the problem?

There will be a bitcoinj 0.16.3 release soon.  See Issue #3410 which has been fixed on both `master` and `release-0.16` branches.

**Comment by ulserGit** [comment]
> > my blockchain download process stopped at block 849137 I'm using BitcoinJ 16.2 from the Maven repository How can I solve the problem?
> 
> There will be a bitcoinj 0.16.3 release soon. See Issue #3410 which has been fixed on both `master` and `release-0.16` branches.

Thank you) how long wait release in to maven repo?

**Comment by msgilligan** [comment]
**Bitcoin Core** uses a 32-byte (256-bit) value using types uint256 and arith_uint256. So maybe we should just serialize the value as 32-bytes.

`SPVBlockStore` is formatted according to the following comment:

```
    // File format:
    //   4 header bytes = "SPVB"
    //   4 cursor bytes, which indicate the offset from the first kb where the next block header should be written.
    //   32 bytes for the hash of the chain head
    //
    // For each header (128 bytes)
    //   32 bytes hash of the header
    //   12 bytes of chain work
    //    4 bytes of height
    //   80 bytes of block header data
    protected static final int FILE_PROLOGUE_BYTES = 1024;
```

So the simplest solution would be to change the header to use `SPV2` and just serialize chainwork size to 32-bytes in `SPVBlockStore`.  That would result in roughly a 16% increase in file/memory size.

**Comment by schildbach** [comment]
I think these days disk and memory space aren't issues any more, even on mobile devices. Bitcoin Wallet uses the default capacity of 10000 blocks, so we're talking about an additional 0.2 MB.

I propose two steps:
- Implement the new format and make sure the SPVBlockStore can handle both formats alike. Adapt relevant tests to test both formats. Once that is done and stable, …
- implement in-place migration of the old format to the new. We already do a similar thing for capacity changes (and this feature was used once in the lifetime of Bitcoin Wallet), so it should be quite easy to do.

If nobody has already started working on this, I'd give it a try.

**Comment by msgilligan** [comment]
@schildbach Sounds like a plan. Should we target this for 0.17 as currently labelled?

**Comment by schildbach** [comment]
> Should we target this for 0.17 as currently labelled?

I'd like to, as we don't know how long our 0.18 cycle will we, and how long apps will then take to adopt 0.18.

**Comment by schildbach** [comment]
Btw. the StoredBlock custom format is also used in both checkpoint formats (binary and textual). We need to adapt them, too.

The write/serialization methods are in `BuildCheckpoints`. The read/deserialization methods are in `CheckpointManager`.

**Comment by schildbach** [comment]
https://github.com/bitcoinj/bitcoinj/pull/3425 implements a first step of a long term solution.

**Comment by regressivetassive** [comment]
I've been exploring different options for managing my BTC transactions and came across an interesting alternative: flash BTC software. I've found that using a reliable flash BTC software can provide a more efficient and secure way to handle transactions. If anyone is interested in learning more, look at that repository (https://github.com/CrypToprexofficial/Free-flash-btc-sender) that provides more information and resources on the topic. Please keep in mind that it's essential to do your own research and due diligence when exploring new tools and methods.

**Comment by schildbach** [comment]
Fixed via #3425, #3430, #3428.
